terraform/provider/resource_teleport_role.go

```diff
@@ -1,61 +1,145 @@
 package provider
 
 import (
-	"github.com/gravitational/teleport-plugins/terraform/tfschema"
+	"context"
 
+	"github.com/gravitational/teleport-plugins/terraform/tfschema"
+	"github.com/gravitational/trace"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/gravitational/teleport/api/types"
+)
+
+const (
+	roleNamePath = "metadata.0.name"
 )
 
+// teleport_role resource definition
 func resourceTeleportRole() *schema.Resource {
 	return &schema.Resource{
-		Create: resourceTeleportRoleUpsert,
-		Read:   resourceTeleportRoleRead,
-		Update: resourceTeleportRoleUpsert,
-		Delete: resourceTeleportRoleDelete,
-		Exists: resourceTeleportRoleExists,
+		CreateContext: resourceRoleCreate,
+		ReadContext:   resourceRoleRead,
+		UpdateContext: resourceRoleUpdate,
+		DeleteContext: resourceRoleDelete,
 
 		Schema: tfschema.SchemaRoleV3(),
 	}
 }
 
-func resourceTeleportRoleUpsert(d *schema.ResourceData, m interface{}) error {
-	// client := m.(*client.Client)
-	name := d.Get("name").(string)
+// resourceRoleCreate creates Teleport role
+func resourceRoleCreate(ctx context.Context, d *schema.ResourceData, m interface{}) diag.Diagnostics {
+	c, err := getClient(m)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
 
-	// roleSpec := services.RoleSpecV3{}
+	name, err := getRoleName(d)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
 
-	// role, err := services.NewRole(name, roleSpec)
-	// if err != nil {
-	// 	return trace.Wrap(err)
-	// }
+	role, err := types.NewRole(name, types.RoleSpecV3{})
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
 
-	// FIXME: Client.CreateRole() is not implemented yet in api/client
-	// err = client.CreateRole(role)
-	// if err != nil {
-	// 	return trace.Wrap(err)
-	// }
+	roleV3, ok := role.(*types.RoleV3)
+	if !ok {
+		return diag.FromErr(trace.Errorf("failed to convert created role to types.RoleV3"))
+	}
 
-	d.SetId(name)
-	return nil
+	err = tfschema.UnmarshalRoleV3(d, roleV3)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
+
+	err = c.UpsertRole(ctx, roleV3)
+	if err != nil {
+		return diag.FromErr(trace.Wrap(err))
+	}
+
+	d.SetId(roleV3.GetName())
+
+	return resourceRoleRead(ctx, d, m)
 }
 
-func resourceTeleportRoleRead(d *schema.ResourceData, m interface{}) error {
-	// client := m.(*auth.Client)
-	// name := d.Get("name").(string)
+// resourceRoleRead reads Teleport role
```

Comment:
What is the purpose of just reading the role? To check its existence? Clarify this in the Godoc please.

Reply:
Yes, it is required by Terraform. Added short description.